README.md
# Windows 11 Drag & Drop to the Taskbar (Fix)

This program fixes the missing "Drag & Drop to the Taskbar" support in Windows 11. In the best case, such a simple and basic functionality will return to Windows 11 in October 2022. For now, we have to use workarounds in the OS that was supposed to "empower our productivity".

**UPDATE:** Windows 11 22H2 (build 22621 and 22622, released in September 2022) restores the drag & drop functionality natively, so this program is not longer needed. To learn how to update your Windows 11, please [visit this page](https://www.microsoft.com/software-download/windows11). You can still use this program if you're running an older version of Windows 11. In case you have troubles removing it, please read [#issue 81](https://github.com/HerMajestyDrMona/Windows11DragAndDropToTaskbarFix/issues/81).

Windows11DragAndDropToTaskbarFix works with the new Windows 11 taskbar and does not require nasty changes like `UndockingDisabled` or restoration of the `classic taskbar`. It doesn't change any system registry keys (except autostart for itself when manually configured) and does not inject any DLLs to other processes, so it's a very portable solution. This program will not interrupt further windows updates as it doesn't change any system files. It's an independent C++ process that does all the job.

### [Download Binary (Windows11DragAndDropToTaskbarFix.exe)](https://github.com/HerMajestyDrMona/Windows11DragAndDropToTaskbarFix/releases/)

### [Install Microsoft Visual C++ 2015-2022 Redistributable x64](https://aka.ms/vs/17/release/vc_redist.x64.exe) 
(in case you're getting the missing `VCRUNTIME140_1.dll` error).

### [Read the Changelog](https://github.com/HerMajestyDrMona/Windows11DragAndDropToTaskbarFix/blob/main/CHANGELOG.md)

### [View the source code (Windows11DragAndDropToTaskbarFix.cpp)](https://github.com/HerMajestyDrMona/Windows11DragAndDropToTaskbarFix/blob/main/Windows11DragAndDropToTaskbarFix/Windows11DragAndDropToTaskbarFix.cpp)


## **How Does It Work?**
<a href="https://github.com/HerMajestyDrMona/Windows11DragAndDropToTaskbarFix/releases/"><img align="right" src="https://user-images.githubusercontent.com/24290461/140068775-473a45f7-1f6a-4b10-8664-e5f8d49440ab.png"></a>

The program detects if you're currently pressing the Left or the Right Mouse Button and determines which icon on the taskbar you hover the mouse pointer on. If the cursor stays in the same area for definied number of milliseconds - it simulates the `Win+T` hotkey and arrow keys in order to restore the intended window, using a quite complicated method. It also supports dropping files to the "Show desktop" button (bottom-right of the screen). The program supports multiple screens, auto startup, and has many configuration options (please read below). Windows11DragAndDropToTaskbarFix has no auto-update function and never connects to the Internet, so you don't need to worry about your privacy.

You can click on the Mona Lisa Megan Fox tray icon to show menu that allows you to `Check for updates`, `Read the changelog`, `Configure`, `Open program folder`, `Restart` and `Quit`.
<br>

## **Custom Configuration: Autostart, Timings, Keys Remapping and more**
It's possible to configure the program by creating a configuration file at the same location as `Windows11DragAndDropToTaskbarFix.exe`. The configuration file should be named: `Windows11DragAndDropToTaskbarFixConfig.txt`. You can automate this process by clicking on the Mona Lisa Megan Fox `Tray icon` and then `Configure...` from the menu. It will create the correct file and open the GitHub page for reference.

Please read a separate article: [CONFIGURATION](https://github.com/HerMajestyDrMona/Windows11DragAndDropToTaskbarFix/blob/main/CONFIGURATION.md) for details.


## **Limitations**
- The program does not support pinning apps to the taskbar using the drag and drop gesture, and this feature is currently not planned. However, you still can pin apps to the taskbar by clicking on their icons with the right mouse button, and then "Show more options" -> "Pin to taskbar" from the list. [Example](https://www.thewindowsclub.com/how-to-pin-any-app-to-the-taskbar-in-windows-11).

- If the program does not work for you, and you're using the custom skin modifications / themes, please add the following line to your [configuration file](https://github.com/HerMajestyDrMona/Windows11DragAndDropToTaskbarFix/blob/main/CONFIGURATION.md):
```
DetectKnownPixelColorsToPreventAccidentalEvents=0
```

- I'm not aware of any serious limitations in the current version. Please check [#issues](https://github.com/HerMajestyDrMona/Windows11DragAndDropToTaskbarFix/issues)


## **How to Start or Quit the Program?**

To use the fix, please simply download the binary file above (or compile it yourself) and **start it as any other program**. It works as an independent C++ process.

To quit this program, please click on the Mona Lisa Megan Fox Tray icon, and select "Quit..." from the menu. You can also **kill** `Windows11DragAndDropToTaskbarFix.exe` or `Windows 11 Drag & Drop to the Taskbar (Fix)` **by the Task Manager**.


## Special Thanks To:
<a href="#special-thanks-to"><img align="top" style="margin-right: 1px" height="70px" src="https://user-images.githubusercontent.com/24290461/140594900-93d603d1-4fbc-4865-a211-fcab1de5507d.png"></a>
<a href="#special-thanks-to"><img align="bottom" style="margin-right: 1px" height="70px" src="https://user-images.githubusercontent.com/24290461/141721061-7fcf595f-da2b-4560-b028-0e317e97a7cb.png"></a>

– **Dr. Mona Lisa** for creating this program. Please also check [Unofficial Patch](https://ee2.eu/patch/) for Empire Earth II.
     <br>
– **Martin Morano** for giving permissions to use his awesome "Mona Lisa Megan Fox" artwork, used as the program icon.
     <br>
– **zhongxia_87887** ([veryicon.com](https://www.veryicon.com/icons/system/gesture-series/drag-drop.html)) for providing the free alternative program icon, modified & enhanced by [**QuestYouCraft**](https://github.com/QuestYouCraft).
     <br>
    <br>

## Before / After Videos

Videos recorded thanks to [@Mohammad Suleman](https://www.ilovefreesoftware.com/author/suleman_ilfs) from [ilovefreesoftware.com](https://www.ilovefreesoftware.com/19/windows/how-to-enable-drag-drop-support-in-windows-11-taskbar.html).

Before Windows11DragAndDropToTaskbarFix | After Windows11DragAndDropToTaskbarFix
:-: | :-:
<video src='https://user-images.githubusercontent.com/24290461/140594641-6975cb11-9a1b-434f-a67f-e3cf339366b7.mp4' width=180/> | <video src='https://user-images.githubusercontent.com/24290461/140594643-4f4f01e5-0382-4781-90d6-eafb1d55f361.mp4' width=180/>


### "VCRUNTIME140_1.dll" was not found error
If you get an error saying that "VCRUNTIME140_1.dll" is missing, please install [Microsoft Visual C++ 2015-2022 Redistributable x64](https://aka.ms/vs/17/release/vc_redist.x64.exe). Normally, these runtime libraries are installed by default on Windows 11. If you get this error then you most likely manually removed them before, using "Add or remove programs".

### **Donation?**

If you love this software so much that you would like to buy me a cola, pizza, or anything else - you can make a donation using the button available on my other project's website: [Click here](https://ee2.eu/overview/#donate). Please don't forget to mention in the donation message that it's for this fix, so I can spend it for other things than multiplayer servers for that game :)


CHANGELOG.md
# Change Log for Windows 11 Drag & Drop to the Taskbar (Fix)
Don't read if you're sensitive to misspellings and grammatical errors.

## [[ver. 2.4.0.0](https://github.com/HerMajestyDrMona/Windows11DragAndDropToTaskbarFix/releases/tag/v.2.4.0.0-release)]  - 2023-10-25, 08:30 CET
**Please remember that this program is most likely not needed if you're using Windows 11 version 22H2 or newer, because they restored the native drag & drop support (possibly thanks to me, because I made them feel ashamed. Just kidding.)**

The program is still usefulf for people who turned UAC off with the `HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Policies\System\EnableLUA` registry key, so I released the new version for them:
- Addressed an issue, where the program stopped working correctly after windows updates that broke `MSTaskSwWClass` window position and coordinates. It solves issues [#98](https://github.com/HerMajestyDrMona/Windows11DragAndDropToTaskbarFix/issues/98), [#99](https://github.com/HerMajestyDrMona/Windows11DragAndDropToTaskbarFix/issues/99). Thanks for the solution idea to user [@Mast3r0mid](https://github.com/Mast3r0mid).
- Improved option: `DetectKnownPixelColorsToPreventAccidentalEvents`:
  - Disabled by default, because it's unpredictable, depending on user's configuration and themes.
  - Added support for Windows 11 23H2 Light and Dark themes.
- Added configuration: `UseLowLevelMousePressProcThread=1`. Please read [CONFIGURATION](https://github.com/HerMajestyDrMona/Windows11DragAndDropToTaskbarFix/blob/main/CONFIGURATION.md) for details.
- Possibly included some other tweaks that I experimented with in 2022/2023 before I got bored of this (useless for me now) program.

## [[ver. 2.3.0.0](https://github.com/HerMajestyDrMona/Windows11DragAndDropToTaskbarFix/releases/tag/v.2.3.0.0-release)]  - 2022-02-15, 17:30 CET
- Added the static build: `Windows11DragAndDropToTaskbarFix_Static.exe`, as requested in issue [#57](https://github.com/HerMajestyDrMona/Windows11DragAndDropToTaskbarFix/issues/57)
- Compiled the program using the `Visual Studio 2022 (v143)` platform toolset.
- Added the new custom configuration variables. Please read [CONFIGURATION](https://github.com/HerMajestyDrMona/Windows11DragAndDropToTaskbarFix/blob/main/CONFIGURATION.md) for details:
  - `FixForBugAfterSleepModeUseOldMethod=0`. `1` = true, `0` = false.
  - `FixForBugAfterSleepModeWindowDisplayTimeMilliseconds=100`. Time in milliseconds. It reduces the time changed in ver. 2.2.0.0 which could make people notice the hotfix window more frequently.
  - `CustomLogFile=`. Custom log file name. If non-empty, it will redirect the program output to a file with the name defined after `=`.
- Improved the `FixForBugAfterSleepMode` function behaviour: it will no longer open the hotfix window when the drag-and-drop gesture is not currently being done. It will also make sure that the mouse pointer is hovered over the taskbar window before opening the hotfix window, in order to minimize the risk of full-screen programs being minimized when the taskbar size bug occurs. It solves issue [#52](https://github.com/HerMajestyDrMona/Windows11DragAndDropToTaskbarFix/issues/52)
- Improved `DetectKnownPixelColorsToPreventAccidentalEvents` behaviour: It will no longer check for correct pixels, if the mouse button was pressed down on a different screen than the one where a file is being drag-and-dropped to. It might also fix issue [#60](https://github.com/HerMajestyDrMona/Windows11DragAndDropToTaskbarFix/issues/60).
- Fixed a bug where the program did not work when the `Primary Mouse Button` was changed to `Right` in Windows 11 Settings -> Bluetooth & Devices -> Mouse. Thanks for the report in discussion [#64](https://github.com/HerMajestyDrMona/Windows11DragAndDropToTaskbarFix/discussions/64).
- Probably some other minor tweaks which I don't remember now.

## [[ver. 2.2.0.0](https://github.com/HerMajestyDrMona/Windows11DragAndDropToTaskbarFix/releases/tag/v.2.2.0.0-release)]  - 2022-01-10, 14:00 CET
- Added `StartThisProgramAsAdministrator=0` configuration option (disabled by default), which allows you to start this program as administrator. It's useful when you often use the Drag & Drop from the other programs that are running as administrator.

Windows11DragAndDropToTaskbarFix/resource.h
//{{NO_DEPENDENCIES}}
// Microsoft Visual C++ generated include file.
// Used by Windows11DragAndDropToTaskbarFix.rc

// Next default values for new objects
// 
#ifdef APSTUDIO_INVOKED
#ifndef APSTUDIO_READONLY_SYMBOLS
#define _APS_NEXT_RESOURCE_VALUE        101
#define _APS_NEXT_COMMAND_VALUE         40001
#define _APS_NEXT_CONTROL_VALUE         1001
#define _APS_NEXT_SYMED_VALUE           101
#endif
#endif

//Icons / Try Icons:
#define IDI_ICON1 101
#define IDI_ICON2 102
#define ID_TRAY_APP_ICON    1001
#define ID_TRAY_EXIT        1002
#define ID_TRAY_SHOW        1003
#define ID_TRAY_RESTART        1004
#define ID_TRAY_CONFIGURE        1005
#define ID_TRAY_CHECKUPDATES        1006
#define ID_TRAY_ABOUT        1007
#define ID_TRAY_CHANGELOG        1008
#define ID_TRAY_OPENPATH        1009
#define ID_TRAY_RESTARTADMIN        1010
#define WM_SYSICON          (WM_USER + 1)



Windows11DragAndDropToTaskbarFix/Windows11DragAndDropToTaskbarFix.h
#pragma once

#include <iostream>
#include <windows.h>
#include <string.h>
#include <chrono>
#include <Shlobj.h>
#include <filesystem>
#include "shobjidl.h"
#include "shlguid.h"
#include "strsafe.h"
#include <set>
#include <psapi.h>
#include <sstream>
#include <fstream>
#include <shlwapi.h>
#include <regex>
#include <thread>
#include <ole2.h>
#include <uiautomation.h>
#include <atlbase.h>
#include "resource.h"

#pragma comment(lib, "Shlwapi.lib")
#pragma comment(lib, "uiautomationcore.lib")
#pragma comment( linker, "/subsystem:windows" )
using namespace std;
using namespace std::chrono;
namespace fs = std::filesystem;
#define DONT_INCLUDE_UNUSED_FUNCTIONS_TO_PREVENT_PSEUDO_ANTIVIRUSES_FROM_THROWING_FALSE_POSITIVES 1

//Basic configuration:
bool AutomaticallyRunThisProgramOnStartup = false;
bool ShowConsoleWindowOnStartup = false;
bool PrintDebugInfo = false;
bool UseFixForBugAfterSleepMode = true;
bool ShowTrayIcon = true;
bool UseTheNewBestMethodEver = true;
bool AutoOpenFirstWindowInBestMethodEver = true;
bool AutoOpenFirstWindowInBestMethodEverLimited = true;
bool AutoOpenPinnedAppsEvenWhenNoWindowActive = false;
bool DetectKnownPixelColorsToPreventAccidentalEvents = false;
bool IgnorePotentiallyUnwantedDragsFromCertainCursorIcons = false;//False since ver 1.10.0.0, because we don't need it with pixels test.
int HowLongSleepBetweenDifferentKeysPressMilliseconds = 20;
int HowLongSleepBetweenTheSameKeysPressMilliseconds = 0;
int HowLongSleepAfterAutoOpenFirstWindowMilliseconds = 100;
int HowLongSleepAfterOpeningPinnedAppMilliseconds = 500;
int HowLongKeepMouseOverAppIconBeforeAutoOpeningMilliseconds = 550;//So it will open after 200+550 (750 ms as in ver. 1.0)
int PreviewWindowChangeDetectionMaxMilliseconds = 1000;//Keep it higher. It's non-blocking time.
bool UseLowLevelMousePressProcThread = true;

bool UseTheNewIconsCountingMethod = true;
bool AutoDetectAndFixIncorrectMSTaskSwWClassWindowSize = false;
int AutoDetectedSingleIconWidthChangeOffsetOnMSTaskSwWClassWindow = 0;
bool AutoDetectedSingleIconWidthChangeOffsetSucceed = false;
int HowLongSleepWhenAutoDetectingMSTaskSwWClassWindowMilliseconds = 1000;

int FixNumberOfIconsOnTaskbarAddOffset = 0;
int HowLongLeftMouseButtonPressedBeforeContinueMilliseconds = 750;
int HowLongKeepMouseOverAppIconBeforeRestoringWindowMilliseconds = 200;//750 before.
int DefaultSleepPeriodInTheLoopMilliseconds = 100;
int SleepPeriodWhenLeftMouseButtonIsPressedInTheLoopMilliseconds = 25;
int SleepPeriodWhenMouseIsOnAppIconInTheLoopMilliseconds = 10;
int DefaultTaskbarIconWidth = 44;
int DefaultTaskbarIconWidth_Now = 44;
int DefaultTaskbarIconHeight = 48;
int DefaultShowDesktopButtonWidth = 20;
int DefaultSingleWindowPreviewThumbnailWidth = 250;
int DefaultSingleWindowPreviewThumbnailHeight = 250;
int FixForBugAfterSleepModeWindowDisplayTimeMilliseconds = 100;
bool FixForBugAfterSleepModeUseOldMethod = false;
string CustomLogFile = "";
bool HaveCustomLogFile = false;

bool UseAlternativeTrayIcon = false;
bool StartThisProgramAsAdministrator = false;//ver 2.2
bool TheProgramIsRunningAsAdministratorRightNow = false;

//bool KeepConsoleWindowVisibleEvenWhenDebugInfoIsDisabled = false;

bool ConfigFileChangeTimeMonitorAllowed = true;
bool ConfigFileChangeTimeMonitorAutoRestart = false;

//Unused (or actually can be used), but not important. Warning, some functions are not longer included in the release
bool UseTheNewWMHOTKEYMethod = false;//Not that reliable method as I thought. False since ver 2.1, too pro blematic.
bool UseTheNewWorkaroundForButtonsElevenPlus = false;//Not needed. Wasted time on it :(
bool CheckIfPinnedAppsWindowsAreVisible = false;//Not needed. Wasted time on it :(
bool DetectIfFileIsCurrentlyDraggedUsingClipboard = true;//Not working for now
int CheckForNewActiveWindowForButtonsElevenPlusMilliseconds = 2500;//Unused by default
int SleepTimeButtonsElevenPlusMilliseconds = 5;//Unused by default
int AnimationLagButtonsElevenPlusMilliseconds = 100;//Unused by default

//Dynamic variables:
wstring ProgramVersion = L"2.4.0.0";
wstring GitHubConfiguration = L"https://github.com/HerMajestyDrMona/Windows11DragAndDropToTaskbarFix/blob/main/CONFIGURATION.md";
wstring GitHubReleases = L"https://github.com/HerMajestyDrMona/Windows11DragAndDropToTaskbarFix/releases";
wstring GitHubAbout = L"https://github.com/HerMajestyDrMona/Windows11DragAndDropToTaskbarFix";
wstring GitHubChangeLog = L"https://github.com/HerMajestyDrMona/Windows11DragAndDropToTaskbarFix/blob/main/CHANGELOG.md";
wstring KnownPixelDetectedCookieFilename = L"W11DADTTPX_KnownPixelDetected.emptyfile";
wstring CurrentExeWorks = L"";
wstring CurrentExeWorksFilenameOnly = L"";
wstring CurrentExeWorksPath = L"";

long long int RestartedCrashedTimes = 0;
wstring RestartExtraWstring = L"";
std::time_t NowSettingsChangeTime;
DWORD CurrentProcessID = 0;
bool InterruptMainThread = false;
bool InterruptMouseWatchdogThread = false;
bool InterruptRestartProgram = false;
bool InterruptRestartProgramRunAs = false;
bool FixForBugAfterSleepModeTestedThisLoop = false;
bool LeftAndRightMouseButtonsAreCurrentlySwapped = false;
bool LeftAndRightMouseButtonsSwappedPreviousStatus = false;

//Virtual Keys Remap:
UINT REMAP_VK_LWIN = VK_LWIN;
UINT REMAP_VK_MENU = VK_MENU;
UINT REMAP_VK_TAB = VK_TAB;
UINT REMAP_VK_LCONTROL = VK_LCONTROL;
UINT REMAP_VK_LSHIFT = VK_LSHIFT;
UINT REMAP_VK_LEFT = VK_LEFT;
UINT REMAP_VK_RIGHT = VK_RIGHT;
UINT REMAP_VK_UP = VK_UP;
UINT REMAP_VK_DOWN = VK_DOWN;
UINT REMAP_VK_RETURN = VK_RETURN;
UINT REMAP_VK_D = 0x44;
UINT REMAP_VK_T = 0x54;
UINT VIRTUAL_REMAP_VK_LWIN = MapVirtualKey(REMAP_VK_LWIN, 0);
UINT VIRTUAL_REMAP_VK_MENU = MapVirtualKey(REMAP_VK_MENU, 0);
UINT VIRTUAL_REMAP_VK_TAB = MapVirtualKey(REMAP_VK_TAB, 0);
UINT VIRTUAL_REMAP_VK_LCONTROL = MapVirtualKey(REMAP_VK_LCONTROL, 0);
UINT VIRTUAL_REMAP_VK_LSHIFT = MapVirtualKey(REMAP_VK_LSHIFT, 0);
UINT VIRTUAL_REMAP_VK_LEFT = MapVirtualKey(REMAP_VK_LEFT, 0);
UINT VIRTUAL_REMAP_VK_RIGHT = MapVirtualKey(REMAP_VK_RIGHT, 0);
UINT VIRTUAL_REMAP_VK_UP = MapVirtualKey(REMAP_VK_UP, 0);
UINT VIRTUAL_REMAP_VK_DOWN = MapVirtualKey(REMAP_VK_DOWN, 0);
UINT VIRTUAL_REMAP_VK_RETURN = MapVirtualKey(REMAP_VK_RETURN, 0);
UINT VIRTUAL_REMAP_VK_D = MapVirtualKey(REMAP_VK_D, 0);
UINT VIRTUAL_REMAP_VK_T = MapVirtualKey(REMAP_VK_T, 0);

//Tray icon Stuff
UINT WM_TASKBAR = 0;
HWND Hwnd;
HMENU Hmenu;
NOTIFYICONDATA notifyIconData;
wchar_t szTIP[64] = L"Windows 11 Drag & Drop to the Taskbar (Fix)";
wchar_t szClassName[] = L"Windows11DragAndDropToTaskbarFix";
wchar_t szTitleName[] = L"Windows 11 Drag & Drop to the Taskbar (Fix)";
wstring ConfigFile = L"Windows11DragAndDropToTaskbarFixConfig.txt";
wstring ConfigFileBase = L"Windows11DragAndDropToTaskbarFixConfig.txt";
wstring ConfigFileAlternative = L"Windows11DragAndDropToTaskbarFixConfig.txt.txt";
wstring ConfigFileAlternative2 = L"Windows11DragAndDropToTaskbarFix.txt";
HINSTANCE hPrevWindow;
HINSTANCE hInstWindow;
int nShowCmdWindow;
HHOOK HandleLowLevelMousePressProc;
bool LeftButtonPressedATM = false;
bool LeftButtonPressedATM_Real = false;
bool RightButtonPressedATM_Real = false;
bool BugMouseButtonNotReallyClicked = false;
short Last_Step_Reached = 0;
std::chrono::milliseconds LastTimeClickedLeftMouseButton = std::chrono::milliseconds(0);
std::chrono::milliseconds LastTimeClickedLeftMouseButton_Real = std::chrono::milliseconds(0);
std::chrono::milliseconds LastTimeClickedRightMouseButton_Real = std::chrono::milliseconds(0);
POINT MouseClickStartPoint;
POINT MouseClickStartPoint_Client;
long long int Current_UniqueID_of_the_click = 0;
bool LLMP_Temporarily_Dont_Update_UniqueID = false;
long long int Last_UniqueID_session_of_Experimental_Workaround = -1;
POINT P;
POINT P_Client;
char Keyboard_Keys_One_to_Zero[10] = { 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39, 0x30 };
const std::vector<std::pair<WPARAM, LPARAM>> New_WM_HOTKEY_Array_LogoWin_CTRL_Num = {
	std::pair<WPARAM, LPARAM> {(WPARAM)0x20F, (LPARAM)0x31000A},
	std::pair<WPARAM, LPARAM> {(WPARAM)0x210, (LPARAM)0x32000A},
	std::pair<WPARAM, LPARAM> {(WPARAM)0x211, (LPARAM)0x33000A},
	std::pair<WPARAM, LPARAM> {(WPARAM)0x212, (LPARAM)0x34000A},
	std::pair<WPARAM, LPARAM> {(WPARAM)0x213, (LPARAM)0x35000A},
	std::pair<WPARAM, LPARAM> {(WPARAM)0x214, (LPARAM)0x36000A},
	std::pair<WPARAM, LPARAM> {(WPARAM)0x215, (LPARAM)0x37000A},
	std::pair<WPARAM, LPARAM> {(WPARAM)0x216, (LPARAM)0x38000A},
	std::pair<WPARAM, LPARAM> {(WPARAM)0x217, (LPARAM)0x39000A},
	std::pair<WPARAM, LPARAM> {(WPARAM)0x218, (LPARAM)0x30000A},
};
bool CurrentlyLeftMouseButtonIsPressed = false;
bool DetectedHWNDsForThisMouseClick = false;
HWND hWndTray = NULL;
HWND hWndTrayNotify = NULL;
HWND hWndRebar = NULL;
HWND hWndMSTaskSwWClass = NULL;
HWND hWndWindowForShowDesktopArea = NULL;
HWND TaskListThumbnailWnd = NULL;
HWND DesktopWindowXamlSourceHwnd = NULL;
HWND hWndStartBTNClass = NULL;
RECT desktop;
HWND hDesktop = NULL;
int ShowDesktopStartPosition = 0;
int TaskbarWindowWidth = 0;
int NumberOfItemsOnTaskbar = 0;
int LastSimulatedHotkeyPressID = -1;
int PreviousHoveredMouseAppID = -1;
std::chrono::milliseconds FirstTimeClickedLeftMouseButton = std::chrono::milliseconds(0);
std::chrono::milliseconds FirstTimeHoveredOverThisAppIcon = std::chrono::milliseconds(0);
std::chrono::milliseconds TimeNow = std::chrono::milliseconds(0);
std::chrono::milliseconds SleepModeFix_Previous_TimeNow = std::chrono::milliseconds(0);
std::chrono::milliseconds LastTimeCheckedForConfigUpdate = std::chrono::milliseconds(0);
time_t LastSettingsChangeTime = 0;
bool CheckedConfigTimeAtLeastOneTime = false;
long long int Previous_UniqueID_of_the_click;
int SleepPeriodNow = DefaultSleepPeriodInTheLoopMilliseconds;
long long Previous_UniqueID_of_the_click_Best_Method_Ever = -1;
int Previous_Button_Number = -1;
int Previous_WindowsScreenSet = -1;
HWND PreviousForegroundWindow;
int JustClickedEnterForBestMethodEver = 0;
RECT rectPreviousTaskListThumbnailWnd;
bool PreviousTaskListThumbnailWndVisible = false;
bool CurrentlyAwaitingForEnterClick = false;
std::chrono::milliseconds AwaitingForEnterClickSince = std::chrono::milliseconds(0);
wchar_t WindowClassName[MAX_PATH];
class windowsHWNDs {
public:
	HWND hWndTray = NULL;
	HWND hWndTrayNotify = NULL;
	HWND hWndRebar = NULL;
	HWND hWndMSTaskSwWClass = NULL;
	HWND TaskListThumbnailWnd = NULL;
	HWND DesktopWindowXamlSourceHwnd = NULL;
	HWND hWndStartBTNClass = NULL;

};
windowsHWNDs PrimaryScreen;
windowsHWNDs CurrentSecondaryScreen;
std::vector<windowsHWNDs> Array_Windows_by_Screen;
HWND TmpWorkerW = NULL;
HWND TmpTaskList = NULL;
int Current_Mouse_Button_Zero_Left_One_Right = -1;
int Previous_Mouse_Button_Zero_Left_One_Right = -1;
std::wstring Current_Button_Name = L"Unknown";
std::wstring Button_Name_Left = L"Left";
std::wstring Button_Name_Right = L"Right";
std::chrono::milliseconds LastTimeFixedTaskbarSizeBug = std::chrono::milliseconds(0);
long long int TaskbarSizeBugDetectedTimes = 0;
SHELLEXECUTEINFO ShExecInfo = { 0 };
bool ConfigMenuDisabled = false;
int ConfigButtonID = 6;
wstring NameAndVer = L"";
wstring NameAndVer2 = L"";
wstring NameRestart = L"";
wstring NameRestartAsAdmin = L"";
MENUITEMINFO	ItemInfo;
bool AllowedCursorIconInThisClick = true;
bool DetectedIconInThisClick = false;
bool DetectedCorrectPixelsInThisClick = false;
int CorrectPixelStatus = 0;
int WindowsScreenSet = 0;//Primary
int PreviousDPI_WindowsScreenSet = -1;
long long int Previous_DPI_UniqueID_of_the_click = -1;
double Current_DPI_Scale_X = 1.0;
double Current_DPI_Scale_Y = 1.0;
bool EVERDetectedCorrectPixels = false;//Always false on program start.
bool CorrectPixelsEverDetectedUsingCookieFile = false;//False on program start

//Functions:
LRESULT CALLBACK WindowProcedure(HWND, UINT, WPARAM, LPARAM);
void minimize();
void restore();
void InitNotifyIconData();
bool dirExists(const std::wstring& dirName_in);
BOOL FileExistsW(LPCWSTR szPath);
BOOL FileExists(LPCTSTR szPath);
void toUpper(basic_string<char>& s);
void toUpper(basic_string<wchar_t>& s);
void toLower(basic_string<char>& s);
void toLower(basic_string<wchar_t>& s);
string Mona_toUpper(string Input);
wstring Mona_toUpperWs(wstring Input);
string Mona_toLower(string Input);
wstring Mona_toLowerWs(wstring Input);
bool is_number(const std::string& s);
void find_and_replace(string& source, string const& find, string const& replace);
void find_and_replace_ws(wstring& source, wstring const& find, wstring const& replace);
bool NewIsConfigLineEqualTo(string ConfigLine, string SearchFor, string SearchForValue);
long long int NewConfigGetIntValueAfter(string ConfigLine, string SearchFor);
double NewConfigGetDoubleValueAfter(string ConfigLine, string SearchFor);
string trololo(string troll);
bool Mona_Remap_Key(string WhichKey, string ChangeTo, bool DebugPrintNow);
void Mona_Load_Configuration(bool DebugPrintNow);
template <typename TP>
std::time_t to_time_t(TP tp);
std::time_t ReturnConfigFileTime(bool OnlyFirstIf);
bool RegistryGetStringValue(std::wstring& valueBuf, HKEY MainKey, const std::wstring& regSubKey, const std::wstring& regValue);
bool RegistrySetStringValue(HKEY MainKey, const std::wstring& regSubKey, const std::wstring& regValue, const std::wstring& stringToSet);
bool RegistryDeleteKeyValue(HKEY MainKey, const std::wstring& regSubKey, const std::wstring& regValue);
void Check_And_Set_Auto_Program_Startup();
static LRESULT CALLBACK LowLevelMousePressProc(int nCode, WPARAM wParam, LPARAM lParam);
DWORD WINAPI MouseClickWatchdogThread(void* data);
void ResetTmpVariablesFull2();
void ResetTmpVariablesFull();
void ResetTmpVariables();
void NewFunctionToKill(bool ReLaunch);
void Simulate_Show_Desktop_Behaviour();
void Simulate_ALT_Plus_TAB_Hotkey(int SleepFor);
bool Should_Auto_Enter_Now_Question_Mark();
void Check_And_Issue_Auto_Enter_Best_Method_Ever(int ButtonID);
void Finally_The_Best_Method_Ever(int ButtonID, int AllButtonsNumber, int WindowsScreenSet);
static BOOL CALLBACK enumChildWindowCallback_SecondaryTrayWnd(HWND hWnd, LPARAM lparam);
static BOOL CALLBACK enumWindowCallback_SecondaryTrayWnd(HWND hWnd, LPARAM lparam);
bool Fix_Taskbar_Size_Bug(HWND Moomintrollen);
void Update_Primary_Screen_Windows_HWNDSs();
DWORD WINAPI ProgramWindowThread(void* data);
void ClickedOpenPathFromTray();
void ClickedChangelogFromTray();
void ClickedAboutFromTray();
void ClickedCheckForUpdatesFromTray();
void ClickedConfigureFromTray();
LRESULT CALLBACK WindowProcedure(HWND hwnd, UINT message, WPARAM wParam, LPARAM lParam);
bool IsCursorIconAllowed();
bool IsCurrentCursorIconStopOrDrag();
void AdvancedSleep();
void Update_Pseudo_DPI_Scale();
int CheckControlPixelsAboveTheMouseOnTaskbar();
bool KnownPixelColors_CookieFileExists();
bool KnownPixelColors_CookieFileCreate();
BOOL IsElevated();
int Return_Current_Taskbar_Window_Width();


#ifndef DONT_INCLUDE_UNUSED_FUNCTIONS_TO_PREVENT_PSEUDO_ANTIVIRUSES_FROM_THROWING_FALSE_POSITIVES
bool AlreadySetAppDataFolders = false;
wstring AppDataFolder = L"";
std::vector<vector<wstring>> Pinned_apps_lnk_to_exe_vector;
//const size_t maxPids = 1024;
const size_t maxPids = 2048;
std::set<std::wstring> processes;
size_t LastBufferSize = 0;
struct handle_data {
	unsigned long process_id;
	HWND best_handle;
};
typedef BOOL(WINAPI* LPQueryFullProcessImageName)(
	HANDLE hProcess, DWORD dwFlags, LPSTR lpExeName, PDWORD lpdwSize);
typedef BOOL(WINAPI* LPQueryFullProcessImageNameW)(
	HANDLE hProcess, DWORD dwFlags, LPWSTR lpExeName, PDWORD lpdwSize);
POINT MouseClickStartPointCopy;
HWND FirstWindowWhereDragged;
bool ItWasDesktopWindow = false;
long long int Unique_ID_of_Click_with_File_Dragged = -1;
long long int Unique_ID_of_Click_with_File_NOT_Dragged = -1;

HRESULT ResolveIt(HWND hwnd, LPCSTR lpszLinkFile, LPWSTR lpszPath, int iPathBufferSize);
void Check_if_Full_Exe_Path_is_currently_running();
vector<wstring> split_ws(const wstring& s, const wstring& delim, const bool keep_empty, std::wstring AddAtTheEnd, std::wstring AddAtTheTop)
inline unsigned int stoui(const std::string& s);
inline unsigned int stouiWs(const std::wstring& s);
BOOL is_main_window(HWND handle);
BOOL CALLBACK enum_windows_callback(HWND handle, LPARAM lParam);
HWND find_main_window(unsigned long process_id);
BOOL IsProcessRunningPID(DWORD pid);
void Check_if_any_window_from_these_PIDs_is_open();
std::wstring GetBinaryValueFromHKCU(const std::wstring& regSubKey, const std::wstring& regValue);
void Check_Pinned_Apps();
HWND Wait_For_The_New_Active_Window(HWND& ActiveWindowBeforeClickingOnAppIcon, wstring DebugStep);
bool Wait_For_Following_Window_To_Become_Foreground_Under_Mouse(HWND& ActiveWindowBeforeClickingOnAppIcon);
void Experimental_Workaround_for_buttons_Eleven_Plus();
bool Detect_if_Clipboard_Has_Dragged_File_Data();
void TestCall();

#endif